### PR TITLE
INTPYTHON-424 add django_mongodb.parse_uri() to configure DATABASES

### PR DESCRIPTION
**Pull request by aclark4life**
Supersedes #194

### CODE REVIEW

**Comment by aclark4life** [comment]
This is working for me with an Atlas URI

Screenshot 2024-11-25 at 9 01 50 PM

File: django_mongodb/utils.py
Comment (timgraham) [inline django_mongodb/utils.py:48]: url -> settings_dict (to match DatabaseWrapper attribut of the same name).

File: django_mongodb/utils.py
Comment (timgraham) [inline django_mongodb/utils.py:34]: Personal preference: write comments without "we".   "so extra..." saves a few words.

File: django_mongodb/__init__.py
Comment (timgraham) [inline django_mongodb/__init__.py:5]: This file should have an `__all__ = ["parse"]` to clarify the public API.

File: django_mongodb/utils.py
Comment (timgraham) [reply django_mongodb/utils.py:51]: It seems to me that kind of goes against the point of parsing the URI in the first place. The expected behavior is unclear if there is some bug in the parsing that causes a mismatch is parsed vs. unparsed (like at https://github.com/mongodb-labs/django-mongodb/blob/36c57187de430ab4313e8db5c3ce8ab2ab569f5b/django_mongodb/base.py#L162-L166 what takes precedence, the values in HOST or the other values). And then if the dbname remains in the URI, what of code like https://github.com/mongodb-labs/django-mongodb/blob/36c57187de430ab4313e8db5c3ce8ab2ab569f5b/django_mongodb/creation.py#L13-L14 that overrides it?

Comment (aclark4life) [reply django_mongodb/utils.py:46]: We cannot pass the raw URI as host because we need to support the Django SETTINGS dictionary API.

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py:49]: I'm not sure what the right thing to do here is but maybe something like https://github.com/jazzband/dj-database-url/blob/358746c41d022be2bab6dc7fa0ac1b98d59069f2/dj_database_url/__init__.py#L176-L189

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py:29]: Maybe! I'm relying on https://github.com/jazzband/dj-database-url/blob/358746c41d022be2bab6dc7fa0ac1b98d59069f2/dj_database_url/__init__.py#L95-L110 for inspiration.

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py]: Done in fb13cbb2e0adeeebf16f4515933451c66fc5c342

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py]: Done in 4ba50ff3540eace436ed71acb531750d1495be90

File: README.md
Comment (Jibola) [inline README.md:124]: Let's have this be the standard rather than the alternative.

File: django_mongodb/utils.py
Comment (Jibola) [inline django_mongodb/utils.py:42]: I was thinking we could do this (since this can be passed to the `MongoClient`)

```suggestion
        host = [f"{node[0]}:{node[1]}" for node in uri["nodelist"]]
        _, port = host[0].split(":")
```

File: django_mongodb/utils.py
Comment (timgraham) [reply django_mongodb/utils.py]: I'm not sure `**kwargs` is the best API design. It allows passing in additional information not in the URI, but it also allows providing information that would override what's in the URI. Like maybe I pass in `OPTIONS={'foo': 'bar'}}` and I don't realize this silently overwrites the options that are in the URI! The Zen of Python says, "There should be one-- and preferably only one --obvious way to do it."  If the user wants to provide a URI, I think they should be required put all the information the URI accepts in there. If some information needs to be provided to DATABASES that doesn't fit in the URI, then we could add a way to do that.

**Review by Jibola — CHANGES_REQUESTED** [review]

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py]: Right, `dj-database-url` does this:

```

def parse(
    url: str,
    engine: Optional[str] = None,
    conn_max_age: Optional[int] = 0,
    conn_health_checks: bool = False,
    disable_server_side_cursors: bool = False,
    ssl_require: bool = False,
    test_options: Optional[Dict[str, Any]] = None,
) -> DBConfig:
```

File: README.md
Comment (aclark4life) [reply README.md]: Done in 0484032244d8f905cdc12c904ed6c308b67fe6b5!

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py]: Actually, if `nodelist > 1` I think we can just pass `nodelist`. Otherwise if `nodelist == 1` then maybe we set host and port. Port is only getting passed outside of the `PORT` setting when replica set.

Comment (aclark4life) [reply django_mongodb/utils.py]: Fixed in 58cdeb0a957b200253712d44992a88c77f597cf7, 🤞

**Comment by aclark4life** [comment]
Via @timgraham we may look to https://github.com/jazzband/dj-database-url/blob/master/tests/test_dj_database_url.py for some testing inspiration.

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py:29]: Back to no kwargs for now until we decide what specific kwargs to support

**Comment by aclark4life** [comment]
@Jibola @timgraham I think this is ready except for:

- What specific kwargs to accept in `parse()`
- What additional tests should be added
- What have I missed if anything

Tests for this are running in Actions.

File: README.md
Comment (timgraham) [inline README.md:106]: Are you going to update the project template? This also needs `import django_mongodb`. Should I use a URI even with a localhost configuration?

File: README.md
Comment (timgraham) [inline README.md:114]: Explain why would I want or need to use the alternative. Add backticks around DATABASES.

File: django_mongodb/utils.py
Comment (timgraham) [inline django_mongodb/utils.py:29]: Please add a docstring.

File: tests/utils_/test_parse.py
Comment (timgraham) [inline tests/utils_/test_parse.py:26]: OPTIONS

File: .pre-commit-config.yaml
Comment (timgraham) [inline .pre-commit-config.yaml:4]: Please move to another PR.

**Review by timgraham — COMMENTED** [review]
Please test with various URIs to exercise the different code paths and the cases when values aren't present.

File: README.md
Comment (aclark4life) [reply README.md:106]: Yes I'll update the project template and add the import statement to the README, and yes to localhost, here's how I'm running the tests:

```
DATABASE_URL = os.environ.get("DATABASE_URL", "mongodb://localhost:27017/djangotests")
```

Comment (timgraham) [reply README.md:106]: In a similar spirit to my comment below, I think promoting a URL as the default (even for something like localhost) is sending people in a more complicated direction in the long run, especially for Django developers coming to MongoDB. It seems pretty simple to parse a URI to the DATABASES format just by looking at it, or as I suggested, use `print(parse_uri(...))` if in doubt.

Comment (aclark4life) [reply README.md]: Agree, I put it back the way it was, which fits the pattern of Django-way, followed by Mongo-way. @Jibola can you roll with this?

Comment (aclark4life) [reply README.md]: Looks like we don't need `DISABLE_SERVER_SIDE_CURSORS` because it is PG specific

Comment (Jibola) [reply README.md]: **TL;DR**: I concede to this and in terms of how we dictate instructions for the a quickstart or target audience, I think that can be broken out of this one. 

The two conventional ways of getting a local MongoDB URI are:
* through [local atlas](https://www.mongodb.com/docs/atlas/cli/current/atlas-cli-deploy-local/)
* through [`mongo-orchestration`](https://github.com/10gen/mongo-orchestration)
 but both of these methods -- in my opinion -- are still roads less traveled. 
 
I really do see the _average_ case being someone with an Atlas URI pointing to instructions on how to get an Atlas URI via the [Getting Started](https://www.mongodb.com/docs/atlas/getting-started/) page.

Because that average case ends up being a more involved cluster, I perceive the template shooting for those cases where no parsing or local db setup is done as the quickest way to get started. More advanced users in the Django and dev space will definitely be able to understand the parsing but true beginners to MongoDB I foresee still being confused.  

Nonetheless, ruminating over this has less to do with the implementation of `parse` and more to do with what gets spit out from our template. Since our template currently supports the `DATABASES` standard, we'll just keep the PR in alignment that way.

**Comment by aclark4life** [comment]
> LGTM, pending a few more tests. @timgraham for the final ✅

OK added a couple more localhost tests.

**Comment by Jibola** [comment]
Looks like there's still some linting issues.

File: README.md
Comment (timgraham) [inline README.md:124]: Still missing the motivation as to why the alternative is needed or should be chosen, as well as docs for the kwargs.

File: README.md
Comment (timgraham) [inline README.md:130]: why use placeholders for user,pass,db, but not host/port?

File: django_mongodb/utils.py
Comment (timgraham) [inline django_mongodb/utils.py:43]: As far as I can think, this would be useful if something is subclassing our backend. Seems unlikely. In any case,  you could simply make `engine="django_mongodb"` the default kwarg if you want to keep it.

**Review by timgraham — COMMENTED** [review]
I thought I left this comment before but I can't find it. I think we should call the function something more than just `parse()`, like `parse_uri()`. Unlike `dj-database-url`, our library has many purposes and therefore many things that could be parsed.

**Comment by aclark4life** [comment]
Thanks for the feedback, @Jibola and @timgraham ! Updates:
- Rename func
- Removed `engine` kwarg
- Added docs

File: django_mongodb/utils.py
Comment (timgraham) [inline django_mongodb/utils.py:45]: This branch appears to be untested (add an `assert False` and no tests fail). It looks like there may be a bug if port is None appending ":None" to the host.

Also, I see no tests for a non-None PORT.

Comment (aclark4life) [reply django_mongodb/utils.py:45]: Good catches, thanks! Fixed in 05df78c27580765c503d2a19eeb791ed271fbc9e and 05ba4453f8a2a71b5afbaf1828e57da9531e58ee

File: tests/utils_/test_parse_uri.py
Comment (timgraham) [inline tests/utils_/test_parse_uri.py:17]: Is the mocking needed? If I comment out the method (even with networking disabled) or change the return value to e.g. cluster1, I see no failures. Also, the [DNS resolving code](https://github.com/mongodb/mongo-python-driver/blob/1b89da4829945359ae99ecfb579104d2d28f7b27/pymongo/uri_parser.py#L557-L559) only seems applicable to srv URLs, but some tests with non-srv URIs are decorated with `@patch("dns.resolver.resolve")`. If any of this is needed, please add some explanatory comments.

File: tests/utils_/test_parse_uri.py
Comment (Jibola) [inline tests/utils_/test_parse_uri.py:30]: Turns out, based on the spec for `NAME` this should end up returning an empty string if it comes back as `None`.
https://docs.djangoproject.com/en/5.1/ref/settings/#name

Comment (Jibola) [reply tests/utils_/test_parse_uri.py:30]: Additionally, in this case, I think it's best to still have the `name` as a passed through kwarg because Atlas URIs do often not provide a database name. I'll provide an example in the code above about how this would work.

File: django_mongodb/utils.py
Comment (Jibola) [inline django_mongodb/utils.py:48]:
```suggestion
        "NAME": name or uri["database"] or '',
```

**Review by Jibola — DISMISSED** [review]

File: tests/utils_/test_parse_uri.py
Comment (aclark4life) [reply tests/utils_/test_parse_uri.py]: If the URI does not contain a database name, then we don't need to add `NAME`. 1519e0f484682f2ad7b89bbb670c0c9c9fac0498

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py]: I would say let's just not add the key 1519e0f484682f2ad7b89bbb670c0c9c9fac0498

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py:47]: This works as-is because `port` is always in the results of PyMongo's `parse_uri` …

File: tests/utils_/test_parse_uri.py
Comment (timgraham) [reply tests/utils_/test_parse_uri.py]: I don't think the None/'' distinction matters. What does it mean to use Django + MongoDB without specifying a database name? I think NAME must be required, although it's not the job of `parse_uri()` to raise an error. Currently, when trying runserver without NAME:
```
Traceback (most recent call last):
  File "/opt/python3.12.0/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/opt/python3.12.0/lib/python3.12/threading.py", line 989, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tim/code/django/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/home/tim/code/django/django/core/management/commands/runserver.py", line 136, in inner_run
    self.check_migrations()
  File "/home/tim/code/django/django/core/management/base.py", line 581, in check_migrations
    executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/executor.py", line 18, in __init__
    self.loader = MigrationLoader(self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/loader.py", line 58, in __init__
    self.build_graph()
  File "/home/tim/code/django/django/db/migrations/loader.py", line 235, in build_graph
    self.applied_migrations = recorder.applied_migrations()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/recorder.py", line 89, in applied_migrations
    if self.has_table():
       ^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/recorder.py", line 64, in has_table
    tables = self.connection.introspection.table_names(cursor)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/introspection.py", line 10, in table_names
    return sorted([x["name"] for x in self.connection.database.list_collections()])
                                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 741 more times]
RecursionError: maximum recursion depth exceeded
```

File: tests/utils_/test_parse_uri.py
Comment (aclark4life) [reply tests/utils_/test_parse_uri.py:17]: Yes, but only on one of the tests, e526ab995b5b6a0b1ed61b7cca05d39ebc2906b1

File: tests/utils_/test_parse_uri.py
Comment (aclark4life) [reply tests/utils_/test_parse_uri.py:28]: That's the bug I found a while back! OK I put it back 56fe609393d8cde55b466f1f6429ce4d12460566

File: tests/utils_/test_parse_uri.py
Comment (timgraham) [reply tests/utils_/test_parse_uri.py:17]: Please add some explanatory comments about the reason mocking is needed.

File: README.md
Comment (aclark4life) [reply README.md]: Errr, 4d155ac084f13246374142b70c8d429230471173

File: tests/utils_/test_parse_uri.py
Comment (aclark4life) [reply tests/utils_/test_parse_uri.py:17]: Done in 6ec2d516255d59ae408daa4d20ea978c8b73ec11

File: README.md
Comment (timgraham) [reply README.md]: Alex, your first intuition is correct. These variables control Django behavior, not the database driver.

Persistent connections are supported:
https://github.com/django/django/blob/ded485464214a3f69b64402b7d82221279f80008/django/db/backends/base/base.py#L249-L250
https://github.com/django/django/blob/ded485464214a3f69b64402b7d82221279f80008/django/db/backends/base/base.py#L615-L617

Health check don't work because we don't use the cursor API: 
https://github.com/django/django/blob/ded485464214a3f69b64402b7d82221279f80008/django/db/backends/base/base.py#L248
(`backends.base.test_base.ConnectionHealthChecksTests` is skipped)

File: tests/utils_/test_parse_uri.py
Comment (timgraham) [reply tests/utils_/test_parse_uri.py:17]: Ok, now I can reproduce the crash by removing  `@patch`. But that seems sufficient... as far as I can tell, `setUp()` is useless. I think it would be slightly more clear if the comment/patch were inside the test method:
```
        with patch("dns.resolver.resolve"):
            settings_dict = parse_uri(URI)
```

File: README.md
Comment (aclark4life) [reply README.md:138]: 0501907c71ed4a1b1ff74a04bb380d20d3cb6784

File: tests/utils_/test_parse_uri.py
Comment (aclark4life) [reply tests/utils_/test_parse_uri.py]: Well played, 1d7e4456af5a8271ccdc6c02b40d3c294619895d, 3549729392a2c0e88e6699e0516980a4b69e332c

File: django_mongodb/utils.py
Comment (aclark4life) [reply django_mongodb/utils.py:45]: @timgraham Any thoughts on this or rather anything for me to do here per @Jibola ?

Comment (timgraham) [reply django_mongodb/utils.py:45]: I would add a test:
```python
    def test_hosts_without_ports(self):
        settings_dict = parse_uri(
            "mongodb://host1.net,host2.net/myDatabase"
        )
        self.assertEqual(settings_dict["HOST"], "host1.net:27017,host2.net:27017")
        self.assertEqual(settings_dict["PORT"], None)
```

Comment (aclark4life) [reply django_mongodb/utils.py:45]: Thanks, 5d6c38e70037744de9782fda3571120ae6368914

File: tests/utils_/test_parse_uri.py
Comment (timgraham) [reply tests/utils_/test_parse_uri.py:21]: Jib said, "I think it's best to still have the name as a passed through kwarg because Atlas URIs do often not provide a database name." By reading the documentation, I hope the user should be able to understand they can append the database name to their URL. Offering a separate keyword increases complexity and makes it possible for them to provide two conflicting values.

Comment (aclark4life) [reply tests/utils_/test_parse_uri.py:21]: `NAME` has to be set in `DATABASES` else Django crashes. Once #196 is merged that will be enforced by `get_connect_params`.  So end users that try to do `parse_uri("mongodb://localhost")` (or something like it, without setting `NAME`) will get an error from `get_connect_params`.

Our `parse_uri` is inspired by `dj-database-url` which is inspired by [The Twelve-Factor App](https://12factor.net/backing-services) which prioritizes what @timgraham was arguing about with `HOST`. `HOST` should contain a `HOST` and nothing else so that:

Via  [Twelve-Factor App](https://12factor.net/backing-services)

> Resources can be attached to and detached from deploys at will. For example, if the app’s database is misbehaving due to a hardware issue, the app’s administrator might spin up a new database server restored from a recent backup. The current production database could be detached, and the new database attached – all without any code changes.

`dj-database-url` set the precedent that additional `kwargs` passed into `parse_uri` are passed to the `DATABASES` dictionary and `NAME` has already been parsed in the URL, so we definitely would not consider adding an additional `kwarg` for `NAME`. As @timgraham says, just direct users to the documentation that says "The database `NAME` is part of the URL" per  [Twelve-Factor App](https://12factor.net/backing-services).

Comment (timgraham) [reply tests/utils_/test_parse_uri.py]: Reading https://www.mongodb.com/docs/manual/reference/connection-string/, one thing that could be confusing is that the example connection URL is: `mongodb+srv://[username:password@]host[/[defaultauthdb][?options]]`.

Does it imply that the authentication database could be different from the database that the user may want to store their Django models in? Reading https://pymongo.readthedocs.io/en/stable/examples/authentication.html#default-database-and-authsource, it looks like `defaultauthdb` can actually be the default database and you can specify a different authentication db using `authSource` in the query string. So I think our implementation is okay.

Comment (aclark4life) [reply tests/utils_/test_parse_uri.py]: Right, I think it's OK too, but playing devil's advocate for @Jibola , Django says `NAME` [can be an empty string](https://docs.djangoproject.com/en/5.1/ref/settings/#name) and if we assume that MongoDB can auto-assign a collection name then we're back to the logistics of supporting an empty string `NAME` which involves an alternative fix for #196 and additional fixes for `parse_uri`. I don't have a strong preference either way and may even lean towards supporting the empty string `NAME` but I'll pass back to @Jibola for additional feedback.

I will note though that even though Django says `NAME` can be an empty string, that appears just a practical matter for Django to leave it to the back end to decide. If your database is SQLite and your `NAME` is empty what happens? I think SQLite picks a file name. If your database is Postgresql and your `NAME` is empty what happens? I think you get an error, but I'm not sure. Same with MySQL. If the MongoDB preference is for `NAME` to be an empty string and there is no error, then I imagine we can get Django to play along with that.

Comment (timgraham) [reply tests/utils_/test_parse_uri.py]: Django's documentation says the the default of NAME is an empty string, but that doesn't mean it's a valid configuration. Some database backends like sqlite3 [raise an error if NAME isn't provided](https://github.com/django/django/blob/5e998d717f7b4220a1728bd49b66ca0162e2a6cb/django/db/backends/sqlite3/base.py#L157-L161), others like have the notion of [a "no db" connection](https://github.com/django/django/blob/5e998d717f7b4220a1728bd49b66ca0162e2a6cb/django/db/backends/postgresql/base.py#L240-L264) (`'NAME': None`) that is of some limited use. For example, when testing on PostgreSQL, "Normally Django will use a connection to the 'postgres' database to avoid running initialization queries against the production."

If there are cases where using Django + MongoDB without a database name is a valid configuration, we need the details. The behavior I see is:
```
>>> db_name = ''
>>> MongoClient(...)[db_name]
...
  File ".../pymongo/synchronous/mongo_client.py", line 1183, in __getitem__
    return database.Database(self, name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../pymongo/synchronous/database.py", line 133, in __init__
    _check_name(name)
  File ".../pymongo/database_shared.py", line 27, in _check_name
    raise InvalidName("database name cannot be the empty string")
pymongo.errors.InvalidName: database name cannot be the empty string
```

Comment (aclark4life) [reply tests/utils_/test_parse_uri.py]: Well played

File: README.md
Comment (timgraham) [inline README.md:113]: Is it correct to say that the scheme is required in HOST only for srv connections? (i.e. `mongodb://` is never required -- as tests seem to indicate?)

Comment (aclark4life) [reply README.md:113]: Yes, and in `parse_uri` I think we only add scheme to host if srv connection.

Comment (timgraham) [reply README.md:113]: I will add: 
```
`HOST` only needs a scheme prefix for SRV connections (`mongodb+srv://`). A
`mongodb://` prefix is never required.
```

**Review by timgraham — APPROVED** [review]

**Pull request merged**
